### PR TITLE
Comprehensive enhancement to the build API

### LINKED ISSUE #98: Does not resolve clojure.set
```
#error {:message "Invalid :refer, var clojure.set/union does not exist in file a3/main.cljs"
```

Demo code:

```clojure
(ns compile.main
  (:require
            [clojure.set :refer [union]]))

(.log js/console
  (union #{1} #{2}))
```

### LINKED ISSUE #244: Ns form require of npm dependencies does not seem to work
Hi Antonio!

Sorry for bombing :bomb:  Just wanted to report something that you might know already.

Basically I tried to compile a namespace using the new syntax:

```
(ns my.util
  (:require [lodash.camelcase :as lcamelcase]))
```

No success. The old-style `(js/require "lodash.camelcase")`, works well.

Just putting it out there for other folks who might come across this problem.

Thanks as usual for your work on both ClojureScript and `lumo`!

### PR DESCRIPTION
**Pull request by anmonteiro**
fixes #244 
fixes #98

### CODE REVIEW

File: src/cljs/bundled/lumo/closure.cljs [hunk 1]
```diff
@@ -311,52 +312,94 @@
...
+                        [(:url this) (:file this)])]
+       (or url file))))
+  (-relative-path
+    ([this] (deps/-relative-path this nil))
+    ([this opts]
+     (let [file (if-let [file-min (and (#{:advanced :simple} (:optimizations opts))
+                                       (:file-min this))]
+                  file-min
+                  (:file this))]
+       (when (and file (not (js/$$LUMO_GLOBALS.path.isAbsolute file)))
```
Comment (arichiardi) [inline src/cljs/bundled/lumo/closure.cljs]: I guess this guy can become just path now.

File: src/cljs/bundled/lumo/closure.cljs [hunk 2]
```diff
@@ -1518,10 +1574,13 @@
                        {:url (js/$$LUMO_GLOBALS.path.resolve out-file)
                         :out-file (.toString out-file)}))]
     (when (and (not js-module?)
-               (or (not (js/$$LUMO_GLOBALS.fs.existsSync out-file))
-                   (and res (util/changed? out-file res))))
+            (or (not (js/$$LUMO_GLOBALS.fs.existsSync out-file))
```
Comment (arichiardi) [inline src/cljs/bundled/lumo/closure.cljs]: This one too can change (minor of course, just marking)

File: src/cljs/bundled/lumo/closure.cljs [hunk 3]
```diff
@@ -1806,51 +2035,102 @@
...
+    (js-transforms js-module opts)
+
+    (symbol? preprocess)
+    (let [ns (namespace preprocess)
+          _ (when (nil? ns)
+              (throw
+                (ex-info (str "Preprocess symbol " preprocess " is not fully qualified")
+                  {:file (:file js-module)
+                   :preprocess preprocess})))
+          preprocess-ns (symbol ns)]
```
Comment (arichiardi) [inline src/cljs/bundled/lumo/closure.cljs]: TIL you can do this :)

File: src/cljs/snapshot/lumo/repl.cljs [hunk 4]
```diff
@@ -168,31 +168,33 @@
 
 (defn- skip-load?
   [name macros?]
-  ((if macros?
-     '#{cljs.core
+  (if macros?
+    ('#{cljs.core
```
Comment (arichiardi) [inline src/cljs/snapshot/lumo/repl.cljs]: A `contains?` would probably make this more readable

Comment (anmonteiro) [reply src/cljs/snapshot/lumo/repl.cljs]: Pr welcome

File: src/cljs/bundled/lumo/compiler.cljs [hunk 5]
```diff
@@ -169,46 +169,54 @@
...
+        env/*compiler*
+        source
+        nil
+        (merge opts {:verbose false
+                     :load (fn [{dep :name :as m} cb]
+                             (if (or (not-empty (get-in cenv [::ana/namespaces dep :defs]))
+                                   (contains? (:js-dependency-index cenv) (name dep))
+                                   (ana/node-module-dep? dep)
+                                   (ana/js-module-exists? (name dep))
+                                   (deps/find-classpath-lib dep))
```
Comment (arichiardi) [inline src/cljs/bundled/lumo/compiler.cljs]: Just curious, was there a problem with the old load?

File: src/cljs/bundled/lumo/util.cljs [hunk 6]
```diff
@@ -111,11 +109,28 @@
...
+    (or (resource? x)
+        (bundled-resource? x))
+    (let [path (.-src x)]
+      (cond-> path
+        windows? normalize-path))
+
+    (jar-resource? x)
+    (str "jar:file:" (.-jarPath x) "!/" (.-src x))
+
+    :else (throw (js/Error. (str "Expected file, url, or string. Got " (pr-str x))))))
```
Comment (arichiardi) [inline src/cljs/bundled/lumo/util.cljs]: Way better!

File: src/cljs/bundled/lumo/compiler.cljs (hunk 5 above)
Comment (anmonteiro) [reply src/cljs/bundled/lumo/compiler.cljs]: `cljs.js/analyze-deps` will try to load stuff that we don't know how to load (e.g. `"module$...$left_pad"`) but it's safe to assume they exist.
same thing that's done [here](https://github.com/clojure/clojurescript/blob/master/src/main/clojure/cljs/analyzer.cljc#L2113)

Comment (arichiardi) [reply src/cljs/bundled/lumo/compiler.cljs]: Oh cool thanks for explaining that

**Comment by arichiardi** [comment]
I willmention the callback on build in the changelog, tomorrow though :smile:

**Pull request merged**